File: 3524-find-x-value-of-array-i/3524-find-x-value-of-array-i.cpp
class Solution {
public:
    vector<long long> resultArray(vector<int>& nums, int k) {
        vector<long long> ans(k, 0);
        vector<long long> dp(k, 0);

        for (int num : nums) {
            vector<long long> newDp(k, 0);
            int mod = num % k;

            // Start a new subarray containing only num
            newDp[mod] = 1;

            // Extend all previous subarrays
            for (int r = 0; r < k; r++) {
                int newMod = (1LL * r * mod) % k;
                newDp[newMod] += dp[r];
            }

            // Add counts to answer
            for (int r = 0; r < k; r++) {
                ans[r] += newDp[r];
            }

            dp = move(newDp);
        }

        return ans;
    }
};

File: 3524-find-x-value-of-array-i/3524-find-x-value-of-array-i.py
from typing import List

class Solution:
    def resultArray(self, nums: List[int], k: int) -> List[int]:
        result = [0] * k
        dp = [0] * k

        for num in nums:
            next_dp = [0] * k

            next_dp[num % k] += 1

            for r in range(k):
                if dp[r] > 0:
                    new_r = (r * num) % k
                    next_dp[new_r] += dp[r]

            dp = next_dp

            for r in range(k):
                result[r] += dp[r]

        return result